harvester_e2e_tests/integration/test_vm_functions.py

```diff
@@ -596,6 +613,287 @@ def test_create_stopped_vm(api_client, stopped_vm, wait_timeout):
+
+        # Verify the VM be hosted expected
+        code, hosts = api_client.hosts.get()
+        cur_res = [(n['metadata']['name'], vm_calc.node_resources(n)['schedulable'])
+                   for n in hosts['data']]
+        schedulables = [(name, res) for (name, res) in cur_res if res[res_type] > expected_res]
+        if not schedulables:
+            # General case
+            assert expected_host == vmi['status']['nodeName'], (
+                f"VM started but not hosted on expected host: {expected_host}"
+            )
+        else:
+            # ???: node's resources released while creating VM, so the VM be hosted on others
+            assert expected_host in [name for name, _ in schedulables], (
+                f"VM started on another host {vmi['status']['nodeName']},"
+                f" and the resource of expected host {expected_host!r} be updated."
+            )
+        # Stop the VM
+        vm_checker.wait_stopped(unique_vm_name)
+
+    def test_update_cpu(
+        self, api_client, ssh_keypair, vm_shell_from_host, vm_checker,
+        stopped_vm
+    ):
+        unique_vm_name, ssh_user = stopped_vm
+        pub_key, pri_key = ssh_keypair
+
+        code, data = api_client.vms.start(unique_vm_name)
+        assert 204 == code, (code, data)
+        vm_got_ips, (code, data) = vm_checker.wait_interfaces(unique_vm_name)
+        assert vm_got_ips, (
+            f"Failed to Start VM({unique_vm_name}) with errors:\n"
+            f"Status: {data.get('status')}\n"
+            f"API Status({code}): {data}"
+        )
+
+        # Login to VM to check current CPU counts
+        vm_ip = next(iface['ipAddress'] for iface in data['status']['interfaces']
+                     if iface['name'] == 'default')
+        code, data = api_client.hosts.get(data['status']['nodeName'])
+        host_ip = next(addr['address'] for addr in data['status']['addresses']
+                       if addr['type'] == 'InternalIP')
+        with vm_shell_from_host(host_ip, vm_ip, ssh_user, pkey=pri_key) as sh:
+            cloud_inited, (out, err) = vm_checker.wait_cloudinit_done(sh)
+            assert cloud_inited, (
+                f"VM {unique_vm_name} Started {vm_checker.wait_timeout} seconds"
+                f", but cloud-init still in {out}"
+            )
+
+            out, err = sh.exec_command("lscpu -e=cpu | tail -1")
+            out = out.strip()
+        assert out.isdigit() and not err, (
+            f"Failed to list cpu amount, output: {out}\nerror: {err}"
+        )
+        new_cpus = (int(out) + 1) + 1  # (shift to 1) + 1
+
+        # Update CPUs
+        code, data = api_client.vms.get(unique_vm_name)
+        assert 200 == code, (code, data)
+        spec = api_client.vms.Spec.from_dict(data)
+        spec.cpu_cores = new_cpus
+        code, data = api_client.vms.update(unique_vm_name, spec)
+        assert 200 == code, (code, data)
+        vm_restarted, ctx = vm_checker.wait_restarted(unique_vm_name)
```

Comment:
about the test logic i mean, for example here.
'restart' is a test step (thus we assert it), but the 'vms.restart' is executed in checker instead of TC self. 
it,s better to decouple responsibility so actions only in TC and checker only do validation.

Reply:
We had tested all VM operations in `TestVMOperations`, so in other test cases, operations would be one of steps, not check points.

The checker is not designed for validation, as I explained earlier, it is used to observe the state change; and for `restart` state, observe the active pod of VMI is different is the current way to confirm the VM is restarted.
so when you look into the `wait_restarted`, it is coupled with the first result of `vm.get_status`.

vm_checker is a fixture to use the VMChecker _instance_ as a helper to observe state change, and VMChecker _class_ isn’t have any responsibility for test case.  The class is designed based on SOLID, SRP for the class is VM and State, methods are State observations.

You can imaging that the class be implemented in harvester_api, and we are just use it to help our testing.